File: src/command/command.registry.ts
import type { BaseCommand } from "./command";

export class CommandRegistry {
  private readonly commands = new Map<string, BaseCommand>();

  register(command: BaseCommand): void {
    this.commands.set(command.name, command);

    for (const alias of command.aliases ?? []) {
      this.commands.set(alias, command);
    }
  }

  get(name: string): BaseCommand | undefined {
    return this.commands.get(name);
  }

  getAll(): BaseCommand[] {
    return [...new Set(this.commands.values())];
  }
}

File: Mythlog/Command/CommandRegistry.cs
using Mythlog.Command.Commands;

namespace Mythlog.Command;

public class CommandRegistry
{
    private readonly Dictionary<string, ICommand> _commands = new();

    public void Register(ICommand command)
    {
        _commands.Add(command.Name, command);
        foreach (var alias in command.Aliases)
        {
            _commands[alias] = command;
        }
    }

    public ICommand? Get(string name)
    {
        return _commands.GetValueOrDefault(name);
    }

    public List<ICommand> GetAll()
    {
        return _commands.Values.Distinct().ToList();
    }
}